### PR TITLE
feat(ranking): add popularAlternativeNames, detect security packages (#941)

### PR DESCRIPTION
**Pull request by MartinKolarik**
1. Adds a new `isSecurityHeld` flag as discussed in #657. I went with a property at the same level as `isDeprecated`(and not `_searchInternal`) since it may be interesting for the end users too.
2. The property is then used for #941 along with some other checks. Popular packages are excluded from the checks since I think those should get the alternative names even if they are deprecated (applies to e.g. [request](https://www-jsdelivr-com-pr-400.onrender.com/?query=request)).

### CODE REVIEW

**Comment by MartinKolarik** [comment]
Note that I focused on the ranking aspect here and didn't add any skipping logic for the security packages - can add that too if you want.

File: src/config.ts [hunk 1]
```diff
@@ -179,6 +181,8 @@ export const config = {
   retrySkipped: ms('1 minute'),
   retryBackoffPow: 3,
   refreshPeriod: ms('2 minutes'),
+  alternativeNamesNpmDownloadsThreshold: 5000,
+  alternativeNamesJsDelivrHitsThreshold: 10000,
```
Comment (Haroenv) [inline src/config.ts:185]: the jsDelivr threshold seems quite high, wouldn't this rule out eg. create-react-app, which only has 950 downloads on jsDelivr

File: src/saveDocs.ts [hunk 2]
```diff
@@ -58,6 +59,18 @@ async function addMetaData(pkg: RawPkg): Promise<FinalPkg> {
...
 
+  const keepAlternativeNames =
+    final.popular ||
+    (!final.isDeprecated &&
+      !final.isSecurityHeld &&
+      (final.downloadsLast30Days >=
+        config.alternativeNamesNpmDownloadsThreshold ||
+        final.jsDelivrHits >= config.alternativeNamesJsDelivrHitsThreshold));
+
+  if (!keepAlternativeNames) {
```
Comment (Haroenv) [inline src/saveDocs.ts:70]:
```suggestion
  const hasFewDownloads =
    final.downloadsLast30Days <= config.alternativeNamesNpmDownloadsThreshold &&
    final.jsDelivrHits <= config.alternativeNamesJsDelivrHitsThreshold;

  const dropAlternativeNames =
    !final.popular &&
    (final.isDeprecated || final.isSecurityHeld || hasFewDownloads);

  if (dropAlternativeNames) {
```

this reads a little simpler to me, what do you think?

**Review by Haroenv — COMMENTED** [review]
interesting approach!

**Comment by Haroenv** [comment]
I'd like to see what impact that has on the index, but not sure how to test it without doing a complete index, do you have any idea?

**Comment by Haroenv** [comment]
Maybe what we could do is instead of removing alternative names completely, we add a new key called popularAlternativeNames in the searchInternal, add the query rule to that instead of alternative names, make sure the original name Is still in popularAlternativeNames (maybe?). That way after a reindex, we can compare the quality between popularAlternativeNames and alternativeNames

**Comment by MartinKolarik** [comment]
> Maybe what we could do is instead of removing alternative names completely, we add a new key called popularAlternativeNames in the searchInternal, add the query rule to that instead of alternative names, make sure the original name Is still in popularAlternativeNames (maybe?). That way after a reindex, we can compare the quality between popularAlternativeNames and alternativeNames

Agree, that's a good idea.

File: src/@types/pkg.ts [hunk 3]
```diff
@@ -48,6 +49,7 @@ export interface RawPkg {
   downloadsLast30Days: number;
   downloadsRatio: number;
   humanDownloadsLast30Days: string;
+  jsDelivrHits: number;
```
Comment (bodinsamuel) [inline src/@types/pkg.ts:52]: indeed good catch

**Review by bodinsamuel — APPROVED** [review]
LGTM
This will require full reindex unfortunately. I need to finish the work on incremental full reindex that was started #819

**Comment by MartinKolarik** [comment]
@bodinsamuel unrelated but do you know what's currently the bottleneck in full reindex? E.g. would it help to run it on a larger VM with higher concurrency?

**Comment by MartinKolarik** [comment]
Also in issues like #930, I was wondering if throwing more resources at it would help as that's something we maybe would be able to do.

**Comment by MartinKolarik** [comment]
> Rate limit (429) from NPM registry
We should get in touch with them to see what can be done

That's interesting as our services query npm a lot too and we don't have this issue. Is that the registry or some other service? (maybe the downloads API?). We previously had a similar problem with GitHub and it the limit was too low-level for them to do anything about it, we solved it by assigning multiple IPs to our VMs and distributing the outgoing requests across those (but that's probably not possible at PaaS).

> We can barely parallelise work

I checked the code and it indeed looks problematic but I think there might be some not too hard solutions. Process N packages at once, and track which are being processed. When a second event for the same package arrives, either put it in an intermediate queue (which has some small size limit) or simply pause until the previous event for the package competes.

**Comment by MartinKolarik** [comment]
> E.g: For package A and package B, if I receive A1, B1, A2, for whatever reason I process B1 first, I can not say "the process is at B1". If the process crashes during this time, I need to be able to say "I processed B2 but not A1 and A2".

I may be missing something but not sure how the concurrency is relevant here - the inserts/updates here should be idempotent so I would just ignore the fact that B1 was processed and say "I'm at whatever was before A1" - which makes this part the same as for serial processing.

> It's a bit awful how simple it seems at first and then realising that it actually requires more complexity to be able to process things correctly.

I'm not surprised at all 😆

**Comment by bodinsamuel** [comment]
> should be idempotent so I would just ignore the fact that B1 was processed and say "I'm at whatever was before A1"

Yeah my explanation was incomplete, let me try again. With this sequence of updates:
```sh
sequence 1: A1 (update A)
sequence 2: B1 (create B)
sequence 3: A2 (update A)
sequence 4: B2 (delete B)
```

We start at `sequence 0`, we process B1 and B2 correctly but A1 is blocking for whatever reason. The process needs to store the progress of the sequence somewhere to be able to start where it left off in case of crash or restart.
- If I store `sequence 4` I skipped 2 updates
- If I store `sequence 0` I redo 4 updates

As you mention, updates are now idempotent (which was not the case before), so it's fine to reprocess until you realise that if you already processed `sequence 4` you deleted the package B. So by reprocessing all sequences you endup actually recreating the package B. (or in other scenario you virtually rollback to a previous version)

Note:
- it's more than 4 sequences, could be hundred or thousand. So could impact a lot of packages.
- we could say we don't care and it's an acceptable downside
- we could add one more check that would get the last update date for a package and skip all updates where `sequence.updateAt < current.updateAt` but it needs to be stored somewhere and Algolia can't be relied upon because search is only eventually consistent.

That's why I would favor having a DB or a queue, to store all updates and redo only the ones that were not processed.
I hope the explanation is bit clearer :p

**Comment by MartinKolarik** [comment]
> So by reprocessing all sequences you endup actually recreating the package B. (or in other scenario you virtually rollback to a previous version)

This seems acceptable because the window for this error is only as big as the concurrency level so very soon you delete B again. You could even say that from an outside observer's perspective, you wouldn't get to `seq 4` before the crash in the first place with serial processing, so the final correct state may still be achieved faster even if it's `B1 -> B2 -> B1 -> B2` instead of `B1 -> B2`

Anyway, an external queue definitely makes sense here, just saying that if an easier solution was more desirable,  it could probably still work.

**Comment by MartinKolarik** [comment]
Changed to `popularAlternativeNames` and `alternativeNames ` are kept as they were.

**Comment by MartinKolarik** [comment]
> Note that I focused on the ranking aspect here and didn't add any skipping logic for the security packages - can add that too if you want.

Added now.

File: src/saveDocs.ts [hunk 4]
```diff
@@ -58,6 +63,18 @@ async function addMetaData(pkg: RawPkg): Promise<FinalPkg> {
...
+  const hasFewDownloads =
+    final.downloadsLast30Days <= config.alternativeNamesNpmDownloadsThreshold &&
+    final.jsDelivrHits <= config.alternativeNamesJsDelivrHitsThreshold;
+
+  const addPopularAlternativeNames =
+    final.popular ||
+    (!final.isDeprecated && !final.isSecurityHeld && !hasFewDownloads);
+
+  if (addPopularAlternativeNames) {
+    final._searchInternal.popularAlternativeNames = [];
```
Comment (Haroenv) [inline src/saveDocs.ts:75]:
```suggestion
    final._searchInternal.popularAlternativeNames = final._searchInternal.alternativeNames;
```

isn't this meant to be something like this?

Comment (Haroenv) [inline src/saveDocs.ts:75]: We also can do the condition before creating final:

```js
const popularAlternativeNames = [];

const hasFewDownloads =
  final.downloadsLast30Days <= config.alternativeNamesNpmDownloadsThreshold &&
  final.jsDelivrHits <= config.alternativeNamesJsDelivrHitsThreshold;

const addPopularAlternativeNames =
  final.popular ||
  (!final.isDeprecated && !final.isSecurityHeld && !hasFewDownloads);

if (addPopularAlternativeNames) {
  popularAlternativeNames.push(...pkg._searchInternal.alternativeNames);
}

const final = {
  // ...
  _searchInternal: {
    ...pkg._searchInternal,
    popularAlternativeNames,
  },
};
```

or in a function:

```js
function s() {
  const popularAlternativeNames = getPopularAlternativeNames(pkg);
  const final = {
    // ...
    _searchInternal: {
      ...pkg._searchInternal,
      popularAlternativeNames,
    },
  };
}

function getPopularAlternativeNames(pkg) {
  const hasFewDownloads =
    pkg.downloadsLast30Days <= config.alternativeNamesNpmDownloadsThreshold &&
    pkg.jsDelivrHits <= config.alternativeNamesJsDelivrHitsThreshold;

  const addPopularAlternativeNames =
    pkg.popular ||
    (!pkg.isDeprecated && !pkg.isSecurityHeld && !hasFewDownloads);

  if (addPopularAlternativeNames) {
    return pkg._searchInternal.alternativeNames;
  }

  return [];
}
```

Comment (MartinKolarik) [reply src/saveDocs.ts:75]: Lol yes, need to update the tests too.

Comment (MartinKolarik) [reply src/saveDocs.ts:75]: We could but then each property is on a different object and some null checks are needed, so it looks like this:

```js
  const popularAlternativeNames: string[] = [];

  const hasFewDownloads =
      (download?.downloadsLast30Days || 0) <= config.alternativeNamesNpmDownloadsThreshold &&
      hit.jsDelivrHits <= config.alternativeNamesJsDelivrHitsThreshold;

  const addPopularAlternativeNames =
      download?.popular ||
      (!pkg.isDeprecated && !pkg.isSecurityHeld && !hasFewDownloads);

  if (addPopularAlternativeNames) {
    popularAlternativeNames.push(...pkg._searchInternal.alternativeNames);
  }
```

Comment (Haroenv) [reply src/saveDocs.ts]: fine as-is then 👍

**Comment by MartinKolarik** [comment]
Then I think it's all done here.

File: src/config.ts [hunk 5]
```diff
@@ -85,14 +87,14 @@ const indexRules: Rule[] = [
     objectID: 'promote-exact',
     description: 'promote exact matches',
     condition: {
-      pattern: '{facet:_searchInternal.alternativeNames}',
+      pattern: '{facet:_searchInternal.popularAlternativeNames}',
```
Comment (Haroenv) [inline src/config.ts:90]: we should change the query rule only after the reindex is completed, otherwise this will automatically change everything.

Comment (MartinKolarik) [reply src/config.ts]: True, reverted everything to `alternativeNames` for now.

**Review by Haroenv — APPROVED** [review]

**Comment by bodinsamuel** [comment]
Well I don't have admin right neither and we do have the the "run fork pr" in circleci, it should run 🤔

**Comment by Haroenv** [comment]
merged in #975
